### Installing the Pipeline
Last Updated:  April 23, 2025

### Running the Pipeline
To run USCMD, you will need the following data:
- a reference human exome file (fasta format)
- two exome reads files (r1 and r2 in fastq.gz format)
- a possorted genome file output by cellranger (bam format)
- a list of barcodes identifying which cells in the above file to look for mutations in (.txt format, a barcode per line)

### How it works (overview):
This method was implemented in ten steps:
(1) Align and Index Exome carries out the alignment of individual exome sequencing reads to the human reference genome to create a donor exome bam file. 

(2) Filter Single Cell Data: split the combined scRNAseq data output by the CellRanger pipeline (named possorted_genome_bam.bam in the CellRanger output folder) into separate files for each of the individual cells.  

(3) Identify Exome SNPs: identify variants based on this individual exome. 

(4) Create Donor-Specific Reference: filters variants from (3) by checking if they are single nucleotide polymorphisms (SNPs), which were then used to create a donor-specific reference in which their SNPs are incorporated into the individual reference. 

(5) Call and Filter Cell Variants: calls somatic mutations within the single cell bam files generated by step 2 based on the donor-specific reference. 

(6) Reformat Annotated Variants: reformats the output from step5. 

(7) UMI corrections: extracts every read and associated metadata from at each mutated position identified, and outputs them as tsv files. 

(8) Score Mutations for single cells: takes in these TSV files and scores the mutations based on multiple quality control metrics and outputs these scores as a CSV file. 

(9) Identify variant type: Use SnpEff to identify the amino acid change associated with each nucleotide mutation. 

(10) Sequencing depth and coverage: Output the exome depth and, sequencing exome coverage (how many base pairs of the exome were sequenced at sufficient depth), and sequencing reads for all the donors to a single CSV file, which can be used to adjust for differences in sequencing depth and coverage among the donors.

### Running it step by step

#### Step 1
Usage `ffdfd`
