
<!--==========================================================================
	FTL macros 
	========================================================================== -->


<!--==========================================================================
	CSS styles 
	========================================================================== -->


<style type="text/css">

body {
	background-color: #dddddd;
}

/* Table styles. */

table {
  border-color: #000;
  border-spacing: 0px;
  border-style: solid;
  border-width: 1px;
  cell-spacing: 0px;
}

.noBorder {
  border-width: 0px;
}

td, th {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 10pt;
  padding: 2px 0.5em;
  white-space: nowrap;
}

td.numeric {
  text-align: right;
}

th {
  background-color: #c0c0c0;
}

th.mainHeader {
  background-color: #808080;
  color: #ffffff;
  text-align: left;
}

th a {
  color: #000080;
  text-decoration: none;
}

th a:visited {
  color: #000080;
}

th a:active, th a:hover {
  color: #800000;
  text-decoration: underline;
}

.toc {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 95%;
}

.note {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 90%;
	text-align: left;
}

.main {
	margin-top: 15px;
	width: 1000px;
	text-align: left;
	background: #ffffff;
	border: 3px solid #777777;
}

.histo {
	table-layout:fixed;
	width:100%;
	border:1px solid #aaa;
	word-wrap:break-word;
}

</style>


<!--==========================================================================
	Index
	========================================================================== -->

<center>
<div class="main">

<center> <h3> SnpEff: Variant analysis </h3> </center>

<div style="margin-left: .5em">
<table class="toc"><tr><td>
	<center><b>Contents</b></center>
	<a href="#summary">Summary</a><br>
	<a href="#changeRateChr"> Variant rate by chromosome</a><br>
	<a href="#changesByType">Variants by type</a><br>
	<a href="#effectsImpact"> Number of variants by impact </a><br>
	<a href="#effectsImpact"> Number of variants by functional class </a><br>
	<a href="#effects"> Number of variants by effect </a><br>
	<a href="#quality">Quality histogram</a><br>
	<a href="#indels">InDel length histogram</a><br>
	<a href="#baseChages">Base variant table</a><br>
	<a href="#tstv">Transition vs transversions (ts/tv)</a><br>
	<a href="#alleleFreq"> Allele frequency </a><br>
	<a href="#alleleCount"> Allele Count </a><br>
	<a href="#codonChanges"> Codon change table </a><br>
	<a href="#aaChanges"> Amino acid change table </a><br>
	<a href="#chrChanges"> Chromosome variants plots </a><br>
	<a href="snpEff_genes.txt"> Details by gene </a><br>
</tr></td></table>
</div>


<!--==========================================================================
	Summary table 
	========================================================================== -->

<hr>
<a name="summary">
<center>
<b>Summary</b><p>

<table border=0>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Genome </b> </td>
		<td> GRCh38.99 </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Date </b> </td>
		<td> 2021-09-30 00:22 </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> SnpEff version </b> </td>
		<td> <pre>SnpEff 5.0e (build 2021-03-09 06:01), by Pablo Cingolani</pre> </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Command line arguments </b> </td>
		<td> <pre>SnpEff  -no-downstream -no-upstream GRCh38.99 
real_cross_sectional_split_first_part_2021_09_26a/BL5522/step5//mutations_NoIntervals/BL5522_CATTATCTCTTGTATC-1_var_FLTR.vcf 
</pre> </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Warnings </b> </td>
		   
		<td bgcolor="#ffff00"> 2 </td>
	</tr>	
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Errors </b> </td>
		   
		<td bgcolor="#ff0000"> 1 </td>
	</tr>	
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of lines (input file) </b> </td>
		<td> 33 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants (before filter) </b> </td>
		<td> 33 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of not variants <br>(i.e. reference equals alternative) </b> </td>
		<td> 0 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants processed <br> (i.e. after filter and non-variants) </b> </td>
		<td> 33 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of known variants <br>(i.e. non-empty ID) </b> </td>
		<td> 
		0
		( 0% ) 
		</td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of multi-allelic VCF entries <br>(i.e. more than two alleles) </b> </td>
		<td> 
		0
		</td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of effects </b> </td>
		<td> 69 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Genome total length </b> </td>
		<td> 63,147,197,749 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Genome effective length </b> </td>
		<td> 2,276,506,285 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Variant rate </b> </td>
		<td> 1 variant every 68,985,038 bases </td>
	</tr>
</table>
<p>
</center>

<!--==========================================================================
	Variant rate by chromosome
	========================================================================== -->

<hr> 
<a name="changeRateChr"> 

<center>
<b> Variants rate details </b><p>

<table border=1>
	<tr><th> Chromosome </th><th> Length </th><th> Variants </th><th> Variants rate </th></tr> 
		<tr>
			<td> 1 </td>
			<td class="numeric"> 248,956,422 </td>
			<td class="numeric"> 4 </td>
			<td class="numeric"> 62,239,105 </td>
		</tr> 
		<tr>
			<td> 2 </td>
			<td class="numeric"> 242,193,529 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 242,193,529 </td>
		</tr> 
		<tr>
			<td> 3 </td>
			<td class="numeric"> 198,295,559 </td>
			<td class="numeric"> 3 </td>
			<td class="numeric"> 66,098,519 </td>
		</tr> 
		<tr>
			<td> 4 </td>
			<td class="numeric"> 190,214,555 </td>
			<td class="numeric"> 5 </td>
			<td class="numeric"> 38,042,911 </td>
		</tr> 
		<tr>
			<td> 5 </td>
			<td class="numeric"> 181,538,259 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 181,538,259 </td>
		</tr> 
		<tr>
			<td> 7 </td>
			<td class="numeric"> 159,345,973 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 159,345,973 </td>
		</tr> 
		<tr>
			<td> 8 </td>
			<td class="numeric"> 145,138,636 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 145,138,636 </td>
		</tr> 
		<tr>
			<td> 9 </td>
			<td class="numeric"> 138,394,717 </td>
			<td class="numeric"> 4 </td>
			<td class="numeric"> 34,598,679 </td>
		</tr> 
		<tr>
			<td> 11 </td>
			<td class="numeric"> 135,086,622 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 135,086,622 </td>
		</tr> 
		<tr>
			<td> 12 </td>
			<td class="numeric"> 133,275,309 </td>
			<td class="numeric"> 2 </td>
			<td class="numeric"> 66,637,654 </td>
		</tr> 
		<tr>
			<td> 14 </td>
			<td class="numeric"> 107,043,718 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 107,043,718 </td>
		</tr> 
		<tr>
			<td> 15 </td>
			<td class="numeric"> 101,991,189 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 101,991,189 </td>
		</tr> 
		<tr>
			<td> 18 </td>
			<td class="numeric"> 80,373,285 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 80,373,285 </td>
		</tr> 
		<tr>
			<td> 19 </td>
			<td class="numeric"> 58,617,616 </td>
			<td class="numeric"> 3 </td>
			<td class="numeric"> 19,539,205 </td>
		</tr> 
		<tr>
			<td> M </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 0 </td>
			<td class="numeric"> 0 </td>
		</tr> 
		<tr>
			<td> X </td>
			<td class="numeric"> 156,040,895 </td>
			<td class="numeric"> 3 </td>
			<td class="numeric"> 52,013,631 </td>
		</tr> 
	<tr>
		<th> Total </th>
		<th class="numeric"> 2,276,506,285 </th>
		<th class="numeric"> 33 </th>
		<th class="numeric"> 68,985,038 </th>
	</tr> 
</table> 
</center>

<!--==========================================================================
	Variants by type
	========================================================================== -->

<hr> 
<a name="changesByType"> 
<center>
<b> Number variants 